File: Problem3.py
#Time Complexity :O(h+k)
#Space Complexity :O(n)
#Did this code successfully run on Leetcode : Yes
#Any problem you faced while coding this : No

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        self.count=k
        self.result=None
        self.helper(root)
        return self.result.val

    def helper(self,root):
        if not root:
            return 
            
        if self.count>0: 
            self.helper(root.left)
        
        self.count-=1
        if self.count==0:
            self.result=root

        if self.count>0: 
            self.helper(root.right)

File: Problem 3.cpp
//Time Complexity : O(n)
// Space Complexity :O(1)
// Did this code successfully run on Leetcode :Yes
// Any problem you faced while coding this :No
class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        if(root==NULL|| root->val==p->val || root->val==q->val)
            return root;
        TreeNode* l = lowestCommonAncestor(root->left, p,q);
        TreeNode* r = lowestCommonAncestor(root->right, p, q);
        
        if(l!=NULL && r!=NULL) return root;
        
        else if(l!=NULL)
            return l;
        else if(r!=NULL)
            return r;
        
        return NULL;
        
    }
};